File: PISC/potentials/Henon_Heiles.py
import numpy as np
from PISC.potentials.base import PES
#from PISC.potentials.henon_heiles_f import henon_heiles as henon_heiles_f

class henon_heiles(PES):
        def __init__(self,lamda,g):
            super(henon_heiles).__init__()
            self.g = g
            self.lamda = lamda  
            self.param_list = [lamda,g]

        def bind(self,ens,rp,pes_fort=False,transf_fort=False):
            super(henon_heiles,self).bind(ens,rp,pes_fort=pes_fort,transf_fort=transf_fort)
 
        def potential(self,q):
            x = q[:,0] 
            y = q[:,1]
            return self.potential_xy(x,y) 
        
        def dpotential(self,q):
            x = q[:,0] 
            y = q[:,1]
            Vx = 4*self.g*x*(x**2 + y**2) + 2*self.lamda*x*y + x
            Vy = 4*self.g*y*(x**2 + y**2) + self.lamda*(x**2 - y**2) + y
            return self.dpot_rearrange([Vx,Vy]) 

        def ddpotential(self,q):
            x = q[:,0] 
            y = q[:,1]
            Vxx = 8*self.g*x**2 + 4*self.g*(x**2 + y**2) + 2*self.lamda*y + 1.0
            Vxy = 2*x*(4*self.g*y + self.lamda)
            Vyx = 2*x*(4*self.g*y + self.lamda)
            Vyy = 8*self.g*y**2 + 4*self.g*(x**2 + y**2) - 2.0*self.lamda*y + 1.0

            return self.ddpot_rearrange([[Vxx,Vxy],[Vxy,Vyy]])

        def potential_xy(self,x,y):
            vh = 0.5*(x**2+y**2)
            vp = self.lamda*(x**2*y - y**3/3.0)
            vs = self.g*(x**2+y**2)**2  
            return (vh + vp + vs)

        def potential_f(self,qcart_f,pot_f):
            return henon_heiles_f.potential_f(qcart_f,self.param_list,pot_f) 
            
        def dpotential_f(self,qcart_f,dpot_f):
            return henon_heiles_f.dpotential_f(qcart_f,self.param_list,dpot_f)

        def ddpotential_f(self,qcart_f,ddpot_f):
            return henon_heiles_f.ddpotential_f(qcart_f,self.param_list,ddpot_f)

File: PISC/potentials/Henon_Heiles.f90
module henon_heiles
        implicit none
        private
        public :: potential, dpotential, ddpotential
        public :: potential_f, dpotential_f, ddpotential_f
        contains
        subroutine potential(q,ndim,param_list,lp,Vq)
                integer, intent(in) :: ndim, lp
                real(8), intent(in) :: q(ndim)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(out) :: Vq
                real(8) :: x,y, lamda, g
                real(8) :: vh,vp,vs
                
                lamda = param_list(1)
                g = param_list(2)

                vh = 0.5*(x**2+y**2)
                vp = lamda*(x**2*y - y**3/3.0)
                vs = g*(x**2+y**2)**2  
                
                x = q(1)
                y = q(2)
                
                Vq = vh + vp + vs

        end subroutine potential

        subroutine dpotential(q,ndim,param_list,lp,dVq)
                integer, intent(in) :: ndim, lp
                real(8), intent(in) :: q(ndim)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: dVq(ndim)
                real(8) :: x,y, lamda, g
                        
                lamda = param_list(1)
                g = param_list(2)

                x = q(1)
                y = q(2)
                
                dVq(1) = 4*g*x*(x**2 + y**2) + 2*lamda*x*y + x
                dVq(2) = 4*g*y*(x**2 + y**2) + lamda*(x**2 - y**2) + y

        end subroutine dpotential

        subroutine ddpotential(q,ndim,param_list,lp,ddVq)
                integer, intent(in) :: ndim, lp
                real(8), intent(in) :: q(ndim)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: ddVq(ndim,ndim)
                real(8) :: x,y, lamda, g
                         
                x = q(1)
                y = q(2)
                
                lamda = param_list(1)
                g = param_list(2)


                ddVq(1,1) = 8*g*x**2 + 4*g*(x**2 + y**2) + 2*lamda*y + 1.0
                ddVq(1,2) = 2*x*(4*g*y + lamda)
                ddVq(2,2) = 8*g*y**2 + 4*g*(x**2 + y**2) - 2.0*lamda*y + 1.0
                ddVq(2,1) = ddVq(1,2)

        end subroutine ddpotential
 
        subroutine potential_f(qcart,N,ndim,nbeads,param_list,lp,pot)
                integer, intent(in) :: N, ndim, nbeads, lp
                real(8), intent(in) :: qcart(nbeads,ndim,N)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: pot(nbeads,N)
                integer :: i,j

                !!$omp parallel do private(i,j) shared(qcart,pot)
                do i = 1, N
                        do j = 1, nbeads
                                call potential(qcart(j,:,i),ndim,param_list,lp,pot(j,i))
                        end do
                end do
                !!$omp end parallel do

        end subroutine potential_f

        subroutine dpotential_f(qcart,N,ndim,nbeads,param_list,lp,dpot_cart)
                integer, intent(in) :: N, ndim, nbeads, lp
                real(8), intent(in) :: qcart(nbeads,ndim,N)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: dpot_cart(nbeads,ndim,N)
                integer :: i,j

                !!$omp parallel do private(i,j) shared(qcart,dpot_cart)
                do i = 1, N
                        do j = 1, nbeads
                                call dpotential(qcart(j,:,i),ndim,param_list,lp,dpot_cart(j,:,i))
                        end do
                end do
                !!$omp end parallel do

        end subroutine dpotential_f
                
        subroutine ddpotential_f(qcart,N,ndim,nbeads,param_list,lp,ddpot_cart)
                integer, intent(in) :: N, ndim, nbeads, lp
                real(8), intent(in) :: qcart(nbeads,ndim,N)
                real(8), intent(in) :: param_list(lp)
                real(8), intent(inout) :: ddpot_cart(nbeads,ndim,nbeads,ndim,N)
                integer :: i,j

                !!$omp parallel do private(i,j) shared(qcart,ddpot_cart)
                do i=1, N
                        do j=1, nbeads
                                call ddpotential(qcart(j,:,i),ndim,param_list,lp,ddpot_cart(j,:,j,:,i))
                        end do
                end do
                !!$omp end parallel do

        end subroutine ddpotential_f
end module henon_heiles
